### PR TITLE
1825 - Moved non-PySpark helpers out of utils/utils.py into polars_utils/utils.py

### PR DESCRIPTION
**Pull request by garypolzin**
## Description
Trello ticket [#1825](https://trello.com/c/QP4DERhV/1825-1-move-non-pyspark-helper-functions-out-of-utils-utilspy)

Moved the non-PySpark helper functions out of utils/utils.py into polars_utils/utils.py, so Fargate/Polars code no longer transitively needs PySpark installed to use them (that file imports pyspark.sql/pydeequ unconditionally at module level). This unblocks the upcoming ons_pd PySpark→Polars conversion and others behind it, which already call several of these helpers directly.

**Changes:**

- **Moved** get_s3_objects_list, read_partial_csv_content, identify_csv_delimiter, is_csv, construct_s3_uri, get_file_directory, construct_destination_path from utils/utils.py → polars_utils/utils.py, adding docstrings and type hints. Left tombstone comments at the old locations.
- **Deleted get_model_name** — confirmed dead code, referenced nowhere outside its own definition and unit test.
- **Fixed a latent cross-platform bug**: construct_s3_uri used os.path.join, which emits backslashes on Windows — broken for S3 URIs. Replaced with an explicit f-string.
- **Updated call sites in the 4 ingest jobs** that use these helpers (ascwds, capacity_tracker, ons_pd, cqc_pir), plus a bonus one-line fix in projects/_01_ingest/cqc_api/utils/utils.py (split_s3_uri import), same defect class.
- **Narrowed the ascwds and cqc_api Fargate Dockerfiles**, which previously COPY'd the entire utils/ tree — including the pyspark-laden utils.py — into images with no pyspark installed. Now copy only the specific subpaths each image actually needs.
- **Tests**: removed 14 obsolete tests from tests/unit/test_utils.py, added 52 pytest-style tests to tests/test_polars_utils/test_utils.py; fixed mock patch targets in 3 ingest job test files (ascwds, capacity_tracker, ons_pd) since the moved functions are now imported directly rather than via utils.

## Testing
- [X] Unit tests passing
- [x] Step Function runs - [ASCWDS](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:856699698263:execution:1825-move-helpers-Ingest-ASCWDS:825c451f-3b72-4c06-b935-12720c9c3866), [ONS](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/statemachines/edit/arn%3Aaws%3Astates%3Aeu-west-2%3A856699698263%3AstateMachine%3A1825-move-helpers-Ingest-ONSPD?referrerPage=StateMachineDetails) and [PIR](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:856699698263:execution:1825-move-helpers-Ingest-CQC-PIR:c6c4a354-eea9-490a-91c6-24a07ec6e211)
**NOTE:** All these failed as they generate the destination path based on the input path. In main, the CSVs are stored in the separate 'raw-data-engineering' bucket but in branch we can't access that. In branch it will always try to save the parquet to where the CSV file is. I think this is ok though, if it's got to the point of saving then it's already hit all these moved functions and found them successfully

## Checklist (delete if not relevant)
- [X] Unit tests added/amended
- [X] Docstrings added/updated
- [X] Updated CHANGELOG
- [X] Code reviewed by AI
- [X] Moved Trello ticket to PR column

## Reviewer checklist
- [ ] PR solves the Trello ticket requirements
- [ ] Outputs appear reasonable and understandable
- [ ] The overall approach is appropriate and not over-engineered
- [ ] No obvious scalability, performance or interdependency concerns
- [ ] Tests appropriately cover the main behaviour/edge cases
- [ ] Naming and structure are broadly understandable and consistent
- [ ] Docstrings are sufficient for future maintenance
- [ ] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: utils/s3_file_utils.py
Comment (RoyPriceCS50) [inline utils/s3_file_utils.py:1]: These are all ingestion utils. Put them in projects/_01_ingest/utils/ingest_utils.py  ?

With the exception of split_s3_uri.

File: projects/_01_ingest/ascwds/jobs/ingest_ascwds_dataset.py [hunk 1]
```diff
@@ -14,6 +14,15 @@
...
 )
+from utils.s3_file_utils import (
+    construct_destination_path,
+    construct_s3_uri,
+    get_s3_objects_list,
+    identify_csv_delimiter,
+    is_csv,
+    read_partial_csv_content,
+    split_s3_uri,
+)
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/jobs/ingest_ascwds_dataset.py]: See my other comment about moving all those ingestion utils to ingestion folder.

This could be iUtils (ingestion utils if agree on moving)

File: projects/_01_ingest/ascwds/tests/jobs/test_ingest_ascwds_dataset.py [hunk 2]
```diff
@@ -26,7 +26,7 @@ def setUp(self):
 
 
 class IngestSingleFileTest(IngestASCWDSDatasetTests):
-    @patch(f"{PATCH_PATH}.utils.construct_destination_path")
+    @patch(f"{PATCH_PATH}.construct_destination_path")
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/tests/jobs/test_ingest_ascwds_dataset.py]: .iUtils.function_name (if you agree to making ingestion_utils a thing.

File: projects/_01_ingest/capacity_tracker/jobs/ingest_capacity_tracker_data.py [hunk 3]
```diff
@@ -6,34 +6,43 @@
...
 from utils import utils
+from utils.s3_file_utils import (
+    construct_destination_path,
+    construct_s3_uri,
+    get_s3_objects_list,
+    identify_csv_delimiter,
+    is_csv,
+    read_partial_csv_content,
+    split_s3_uri,
+)
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/capacity_tracker/jobs/ingest_capacity_tracker_data.py]: This could be iUtils (ingestion utils if agree on moving).

File: projects/_01_ingest/capacity_tracker/tests/test_ingest_capacity_tracker_data.py [hunk 4]
```diff
@@ -43,8 +43,8 @@ def setUp(self):
             NonResData.expected_rows, NonResSchema.sample_schema
         )
 
-    @patch(f"{PATCH_PATH}.utils.read_partial_csv_content")
-    @patch(f"{PATCH_PATH}.utils.get_s3_objects_list")
+    @patch(f"{PATCH_PATH}.read_partial_csv_content")
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/capacity_tracker/tests/test_ingest_capacity_tracker_data.py]: .iUtils.function_name (if you agree to making ingestion_utils a thing.

File: projects/_01_ingest/ons_pd/jobs/ingest_ons_data.py [hunk 5]
```diff
@@ -4,34 +4,43 @@
...
 from utils import utils
+from utils.s3_file_utils import (
+    construct_destination_path,
+    construct_s3_uri,
+    get_s3_objects_list,
+    identify_csv_delimiter,
+    is_csv,
+    read_partial_csv_content,
+    split_s3_uri,
+)
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ons_pd/jobs/ingest_ons_data.py]: Earlier comment about making iUtils a thing.

File: projects/_01_ingest/cqc_pir/jobs/ingest_cqc_pir_data.py [hunk 6]
```diff
@@ -6,26 +6,33 @@
 
 from schemas.cqc_pir_schema import PIR_SCHEMA
 from utils import utils
+from utils.s3_file_utils import (
+    construct_destination_path,
+    construct_s3_uri,
+    get_s3_objects_list,
+    is_csv,
+    split_s3_uri,
+)
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/cqc_pir/jobs/ingest_cqc_pir_data.py]: Earlier comment about making iUtils a thing.

File: projects/_01_ingest/ascwds/fargate/Dockerfile
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/fargate/Dockerfile:1]: This requires  COPY utils/s3_file_utils.py utils/s3_file_utils.py

Otherwise when clean_ascwds_workplace and other polars stuff runs later on, it'll try to import polars_utils which requires utils.s3_file_utils

**Review by RoyPriceCS50 — COMMENTED** [review]
One comment on the Dockerfile which will cause problems. Other comments are mostly moving stuff to an ingestion_utils script.

File: utils/s3_file_utils.py
Comment (RoyPriceCS50) [reply utils/s3_file_utils.py:1]: Leave it where it is. 

It's used often in ingestion, but also once in polars_utils > utils which is outside project folder.
It's not really "polars_utils" because it's used by spark stuff too.

File: projects/_01_ingest/ascwds/fargate/Dockerfile
Comment (garypolzin) [reply projects/_01_ingest/ascwds/fargate/Dockerfile:1]: good spot!

**Review by RoyPriceCS50 — APPROVED** [review]
Look good.
